### PR TITLE
#1456 Refactor collie client to follow fluent pattern

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #1456)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

**Comment by jackyoh** [comment]
明天 Jenkins server 打開， 我會執行 retry

**Comment by jackyoh** [comment]
retry

File: ohara-common/src/main/java/com/island/ohara/common/util/CommonUtils.java [hunk 1]
```diff
@@ -378,28 +378,81 @@ public static int requireConnectionPort(int value) {
...
+  /**
+   * check the port to which you prepare to bind. The port must be bigger than or equal with zero
+   * and small than 65536. The zero is legal since OS will assign a random port to you.
+   *
+   * @param value port number
+   * @return legal port
+   */
+  public static int requireBindPort(int value) {
+    if (value < 0 || value > 65535)
+      throw new IllegalArgumentException("the legal port range is 1 - 65535, actual:" + value);
```
Comment (saivirtue) [inline ohara-common/src/main/java/com/island/ohara/common/util/CommonUtils.java]: since zero is ok, maybe the legal range is "0-65535" ?

Comment (saivirtue) [reply ohara-common/src/main/java/com/island/ohara/common/util/CommonUtils.java]: Does this duplicated to `requireConnectionPort` ?

File: ohara-agent/src/main/scala/com/island/ohara/agent/NodeCollie.scala [hunk 2]
```diff
@@ -27,13 +27,13 @@ import scala.concurrent.{ExecutionContext, Future}
...
-  def nodes(names: Seq[String])(implicit executionContext: ExecutionContext): Future[Seq[Node]] =
-    Future.traverse(names)(node)
+  def nodes(names: Set[String])(implicit executionContext: ExecutionContext): Future[Seq[Node]] =
+    Future.traverse(names)(node).map(_.toSeq)
   def exist(name: String)(implicit executionContext: ExecutionContext): Future[Boolean] =
     node(name).map(_ => true).recover {
       case _: Throwable => false
     }
-  def exist(names: Seq[String])(implicit executionContext: ExecutionContext): Future[Boolean] =
+  def exist(names: Set[String])(implicit executionContext: ExecutionContext): Future[Boolean] =
```
[the remark this answers is not in the document]
Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/NodeCollie.scala]: which “parameters”? The objects stored in node collie are always distinct.

[the remark this answers is not in the document]
Comment (saivirtue) [reply ohara-agent/src/main/scala/com/island/ohara/agent/NodeCollie.scala]: > which “parameters”? 

(names)

> The objects stored in node collie are always distinct.

got it. So this change is a hint that all the objects in this class should be distinct.

File: ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala [hunk 3]
```diff
@@ -78,6 +77,10 @@ object StreamCollie {
     private[this] var toTopics: Seq[String] = Seq.empty
     private[this] var jmxPort: Int = CommonUtils.availablePort()
 
+    override protected def doCopy(clusterInfo: StreamClusterInfo): Unit = {
+      // TODO: do nothing ?
```
[the remark this answers is not in the document]
Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/StreamCollie.scala]: StreamClusterInfo does not carry enough info to be copied...But it is ok now as we don't support to add node for stream application.

I will add enough comment for this do-nothing

**Comment by chia7712** [comment]
unrelated error

**Pull request merged**